src/item_collaborative_filtering.py

```diff
@@ -23,53 +23,98 @@ def __generate_item_pairs(self, df, item):
         :return: list of tuples, length=df['item'].nunique()
         """
         if item is not None:
-            return [(item, paired_item) for paired_item in df[self.item_column].unique() if paired_item!=item]
+            return [(item, paired_item) for paired_item in df[self.item_column].unique() if paired_item != item]
         else:
             return [(item, paired_item) for
                     item, paired_item in permutations(df[self.item_column].unique(), 2)]
 
+    def __item_users(self, df):
+        """
+        Looks up the users that interacted (e.g. watched, purchased) with an item
+        :param df: dataframe with columns [user_id, item_id]
+        :return: dictionary {item: set(user1, user2, ...)}
+        """
+
+        item_users = df.groupby(
+            self.item_column
+        ).agg(
+            {
+                self.user_column: lambda x: set(x)
+            }
+        ).to_dict()[self.user_column]
 
-    def __count_common_item_pair_users(self, df, item_pair):
+        self.item_users = item_users
+
+    def __count_common_item_pair_users(self, item_pair):
         """
         Computes the number of users that interacted (e.g. watched, purchased) with BOTH items in a pair
-        :param df: dataframe with columns [user_id, item_id]
         :param item_pair: tuple(item1, item2)
         :return: int, number of users in common for the item pair (item1, item2)
         """
         item1, item2 = item_pair
-        item1_users = set(df.loc[df[self.item_column]==item1, self.user_column].unique())
-        item2_users = set(df.loc[df[self.item_column]==item2, self.user_column].unique())
+        try:
+            item1_users = self.item_users[item1]
+            item2_users = self.item_users[item2]
+            common_users_count = len(item1_users.intersection(item2_users))
+            return item_pair, common_users_count
+        except AttributeError:
+            "Extract item users first, using __item_users()"
+
+    def __item_probabilities(self, df):
+        """
+        Computes the probability of a user interacting with an item (e.g. watching, purchasing, liking)
+        :param df:  dataframe with columns [user_id, item_id]
+        :return: dictionary with key=item_id, value = interaction probability (float)
+        """
+        item_probabilities = (
+                df.groupby(
+                    self.item_column
+                ).agg(
+                    {self.user_column: 'nunique'}
+                )/df[self.user_column].nunique()
+        ).to_dict()[self.user_column]
```

Comment:
Not very important but the parsing style seems very different here from lines 38-44 for example :)

Reply:
modified :)